File: _scripts/update.rb
#!/usr/bin/env ruby

require 'nokogiri'
require 'open-uri'
require 'net/http'
require 'openssl'

# read url and ignore SSL verify
def read_url(url)
  uri = URI(url)
  response = nil
  Net::HTTP.start(
      uri.host, uri.port,
      :use_ssl => uri.scheme == 'https',
      :verify_mode => OpenSSL::SSL::VERIFY_NONE) do |https|
    request  = Net::HTTP::Get.new uri.request_uri
    response = https.request request
  end
  response.body
end

def is_tg_link(link)
  return false unless link.text =~ /喷嚏图卦\s?\d+/
  return false unless link['href'].include? 'more.asp?name=xilei&id='

  true
end

base_url  = 'https://www.dapenti.com/blog'
list_url  = "#{base_url}/index.asp"

list_html = Nokogiri::HTML(read_url(list_url), nil, 'gbk')
links = list_html.css 'a'
links.each do |link|
  if is_tg_link(link)

    # parse date
    date_str = link.text[/2\d+/]
    year, month, day = [date_str[0,4], date_str[4..-3], date_str[-2..-1]]
    month = "0#{month}" if month.length < 2

    # post file path
    file = "#{__dir__}/../_posts/#{year}/#{year}-#{month}-#{day}-#{year}#{month}#{day}.md"

    unless File.exist? file

      puts "Downloading #{link['title']}."

      url       = "#{base_url}/#{link['href']}"
      article   = Nokogiri::HTML(read_url(url), nil, 'gbk')
      date_time = article.css('div[align="right"] span.oblog_text').first.text[/[\d-]+ [\d:]+/]

      content   = "---\n"
      content   += "layout: post\n"
      content   += "title: \"#{link['title']}\"\n"
      content   += "date: #{date_time}\n"
      content   += "link: #{url.gsub('https', 'http')}\n"
      content   += "---\n\n"
      content   += article.css('div.oblog_text').first.to_s.encode('utf-8')

      File.write file, content
    end

  end
end

File: update.py
import requests
import re
from os import path
from bs4 import BeautifulSoup

BASE_PATH = path.dirname(__file__)
BASE_URL = 'https://www.dapenti.com/blog/'
LIST_URL = '%sblog.asp?name=xilei&subjectid=70' % BASE_URL
ARTICLE_URL_PATTERN = 'more\.asp\?name=xilei&id=\d+'
ARTICLE_TITLE_PATTERN = '喷嚏图卦\s*(\d{8})'


def get_page(url):
    response = requests.get(url)
    if response.status_code != 200:
        raise Exception('Failed to fetch page: %s' % url)
    return BeautifulSoup(response.content, 'lxml', from_encoding='gb2312')


def find_articles():
    list_page = get_page(LIST_URL)

    links = list_page.find('td', class_='oblog_t_2').find_all('a')
    links = [link for link in links if re.match(ARTICLE_URL_PATTERN, link['href']) and re.search(ARTICLE_TITLE_PATTERN, link.text)]

    articles = []
    for link in links:
        found = re.search(ARTICLE_TITLE_PATTERN, link.text)
        if not found:
            continue

        date = found.group(1)
        date = [date[0:4], date[4:6], date[6:]]

        post_file = path.join(BASE_PATH, 'jekyll', '_posts', date[0], '%s-%s-%s-%s.html' % (date[0], date[1], date[2], ''.join(date)))

        if not path.isfile(post_file):
            articles.append({
                'title': link.text,
                'url': BASE_URL + link['href'],
                'date': date,
                'file': post_file
            })

    return articles


def collect_article(article):
    html = get_page(article['url'])
    content = html.find('div', class_='oblog_text')

    publish_time = '00:00:00'
    for segment in html.find_all('span', class_='oblog_text'):
        matches = re.search('发布于\s*(\d+)-(\d+)-(\d+)\s+([\d:]+)', segment.decode())
        if matches:
            publish_time = matches.group(4)
            break

    front_matters = [
        '---',
        'layout: post',
        'title: %s' % article['title'],
        'date: %s %s' % ('-'.join(article['date']), publish_time),
        'link: %s' % article['url'],
        '---'
    ]

    post_contents = '\n'.join(front_matters) + '\n\n' + content.decode().replace('\r\n', '\n')

    with open(article['file'], 'w') as f:
        f.write(post_contents)


def main():
    articles = find_articles()
    for article in articles:
        collect_article(article)


if __name__ == '__main__':
    main()
